core/embed/rust/src/ui/display/toif.rs

```diff
@@ -0,0 +1,183 @@
+
+    for py in area.y0..area.y1 {
+        for px in area.x0..area.x1 {
+            let p = Point::new(px, py);
+            let x = p.x - area.x0;
+
+            if clamped.contains(p) {
+                if x % 2 == 0 {
+                    unwrap!(ctx.uncompress(&mut dest), "Decompression failed");
+                    pixeldata(colortable[(dest[0] & 0xF) as usize]);
+                } else {
+                    pixeldata(colortable[(dest[0] >> 4) as usize]);
+                }
+            } else if x % 2 == 0 {
+                //continue unzipping but dont write to display
+                unwrap!(ctx.uncompress(&mut dest), "Decompression failed");
+            }
+        }
+    }
+
+    pixeldata_dirty();
+}
+
+/// Holding toif data and allowing it to draw itself.
+#[derive(PartialEq, Eq, Clone, Copy)]
+pub struct Toif {
+    data: &'static [u8],
+    name: &'static str, // useful for debugging purposes.
+    size: Offset,
+    format: ToifFormat,
+}
+
+impl Toif {
+    pub fn new(named_toif: NamedToif) -> Self {
+        let info = unwrap!(toif_info(named_toif.0));
+        Self {
+            data: named_toif.0[TOIF_HEADER_LENGTH..].as_ref(),
+            name: named_toif.1,
+            size: info.0,
+            format: info.1,
+        }
+    }
+
+    pub fn from_slice(data: &'static [u8]) -> Self {
+        let named_toif = NamedToif(data, "");
+        Self::new(named_toif)
+    }
+
+    pub fn ensure(&self, format: ToifFormat) {
+        assert_eq!(self.format, format);
+    }
+
+    pub fn format(&self) -> ToifFormat {
+        self.format
+    }
+
+    pub fn name(&self) -> &'static str {
+        self.name
+    }
+}
+
+pub trait ToifAccessors {
+    fn toif(&self) -> Toif;
+
+    fn data(&self) -> &[u8] {
+        self.toif().data
+    }
+
+    fn format(&self) -> ToifFormat {
+        self.toif().format
+    }
+
+    fn width(&self) -> i16 {
+        self.toif().size.x
+    }
+
+    fn height(&self) -> i16 {
+        self.toif().size.y
+    }
+
+    fn size(&self) -> Offset {
+        self.toif().size
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, Copy)]
+pub struct Icon {
+    toif: Toif,
+    area: Rect,
+    precise_x: bool,
+}
+
+impl Icon {
+    pub fn new(named_toif: NamedToif) -> Self {
+        let toif = Toif::new(named_toif);
+        ensure!(toif.format() == ToifFormat::GrayScaleEH, toif.name());
+        Self {
+            toif,
+            area: Rect::zero(),
+            precise_x: false,
+        }
+    }
+
+    pub fn with_precise_x(mut self) -> Self {
+        self.precise_x = true;
+        self
+    }
+
+    /// Display the icon with baseline Point, aligned according to the
+    /// `alignment` argument.
+    pub fn draw(&self, baseline: Point, alignment: Alignment2D, fg_color: Color, bg_color: Color) {
+        let r = Rect::snap(baseline, self.size(), alignment);
+
+        if self.precise_x {
+            icon_precise(self, r.center(), fg_color, bg_color);
+        } else {
+            icon(
```

Comment:
is there actually any advantage of keeping this? could we always use `icon_precise` (and thus get rid of `precise_x` and redefine `Icon` as `struct Icon(Toif)` ?)

Reply:
there are performance benefits of using the C functions (it does less boundaries checking and when DMA2D is enabled it uses that). 

though i think that we could easily live with more boundaries checking and i could implement the DMA2D variant in rust, and when bootloaders get merged we could drop the C functions

Reply:
implemented: https://github.com/trezor/trezor-firmware/commit/69294044a1e7b7239a80dad1e9045d7d561826c7

dma2d has its own limits (mainly can't render odd number of pixels) so in relatively rare situations this is needed, this defaults to non-dma2d rendering.

this happens when part of the icon is out of screen, and rendering starts (or ends) on odd coordinate -> this leads to odd number of pixels in line not shown, therefore odd number of total pixels is rendered since the grayscale toif is always even-wide.

Reply:
removed the `precise_x` https://github.com/trezor/trezor-firmware/commit/8f4ec36f6c8163aa2ca9cb295a51992ecf334404 - didn't do the `struct Icon(Toif)` change yet - what's the benefit there? getting the toif via icon.0 seems less readable to me

removed the dma2d function and left only the precise rust rendering functiton: https://github.com/trezor/trezor-firmware/commit/d227728be3f4e7202784f12e4515afb0fd076223

there is large UI diff again https://github.com/trezor/trezor-firmware/commit/aeb475117b6ba3ae452713f8e312a8449e28e067 because making all icons render on precise location shifted 
- notification icon again
- scrollbar
- cancel button icon in pin keyboard
- checklist icon

![scroll](https://user-images.githubusercontent.com/37884576/214104820-e11ac8ca-951a-40cc-9031-f5cccc809bb0.png) ![scroll2](https://user-images.githubusercontent.com/37884576/214104828-51caa1a1-4f15-4f9c-90ec-f42ff587054f.png) ![scroll_diff](https://user-images.githubusercontent.com/37884576/214104837-e4189dcc-a953-43b5-b23c-3a8b8c4710c8.png)

![cancel](https://user-images.githubusercontent.com/37884576/214104940-587d19ab-18b1-4c66-915c-68cfb1f730fd.png) ![cancel_after](https://user-images.githubusercontent.com/37884576/214104944-ade9293f-6f82-43e2-9d47-be1ba6b47263.png) ![diff](https://user-images.githubusercontent.com/37884576/214104946-92087f78-13aa-4a0b-8ec1-1cc603132c79.png)


![checklist](https://user-images.githubusercontent.com/37884576/214104982-0b91b4e6-7f5d-40f2-beae-23d5a41ae285.png) ![checklist1](https://user-images.githubusercontent.com/37884576/214104979-4d86ac3b-ab95-4477-8850-55b1dda4bd99.png) ![checklist_diff](https://user-images.githubusercontent.com/37884576/214104977-34bb1040-60d2-4e5a-85db-6e8ff7cc7118.png)

Reply:
> didn't do the struct Icon(Toif) change yet - what's the benefit there? getting the toif via icon.0 seems less readable to me

not much, it just makes it a little clearer that the `Icon` is just a newtype for a `Toif`. I'm ok with keeping the item named.